crates/utils/src/lib.rs

```diff
@@ -1,10 +1,51 @@
 use log::{info, warn};
-use std::{path::Path, sync::Arc};
+use std::{os::fd::FromRawFd, path::Path, sync::Arc};
 
 use vhost::vhost_user::{self, Listener};
 use vhost_user_backend::{VhostUserBackend, VhostUserDaemon, VringT};
 use vm_memory::{mmap::NewBitmap, GuestMemoryAtomic, GuestMemoryMmap};
 
+#[cfg(not(feature = "systemd"))]
+fn create_listener<P: AsRef<Path>>(socket_path: P) -> vhost_user::Result<Listener> {
+    Listener::new(socket_path, true)
+}
+
+#[cfg(feature = "systemd")]
+fn create_listener<P: AsRef<Path>>(socket_path: P) -> vhost_user::Result<Listener> {
+    use log::error;
+
+    let fds = systemd::daemon::listen_fds(false).unwrap();
+    let suitable_socket_fd = fds.iter().find(|&fd| {
+        let socket_path = Some(
+            socket_path
+                .as_ref()
+                .to_str()
+                .expect("socket_path should be valid utf-8 string"),
+        );
+
+        systemd::daemon::is_socket_unix(
+            fd,
+            Some(systemd::daemon::SocketType::Stream),
+            systemd::daemon::Listening::IsListening,
+            socket_path,
+        )
+        .unwrap()
+    });
+
+    if let Some(fd) = suitable_socket_fd {
+        unsafe { Ok(Listener::from_raw_fd(fd)) }
```

Comment:
this is broken since we need to grantee that this only happens once (by clearing the environment)